src/inmanta/agent/code_manager.py

```diff
@@ -63,71 +62,66 @@ async def get_pip_config(self, environment: uuid.UUID, version: int) -> PipConfi
-                ),
-            )
-            return resource_install_spec
-        else:
-            raise CouldNotResolveCode(resource_type, version, str(result.get_result()))
-
-    async def get_code(
-        self, environment: uuid.UUID, version: int, resource_types: Collection[ResourceType]
-    ) -> tuple[Collection[ResourceInstallSpec], executor.FailedResources]:
+    async def get_code(self, environment: uuid.UUID, model_version: int, agent_name: str) -> Collection[ModuleInstallSpec]:
         """
         Get the collection of installation specifications (i.e. pip config, python package dependencies,
-        Inmanta modules sources) required to deploy a given version for the provided resource types.
-
-        Expects at least one resource type.
+        Inmanta modules sources) required to deploy resources on a given agent for a given configuration
+        model version.
 
         :return: Tuple of:
             - collection of ResourceInstallSpec for resource_types with valid handler code and pip config
             - set of invalid resource_types (no handler code and/or invalid pip config)
         """
-        if not resource_types:
-            raise ValueError(f"{self.__class__.__name__}.get_code() expects at least one resource type")
-
-        resource_install_specs: list[ResourceInstallSpec] = []
-        invalid_resources: executor.FailedResources = {}
-        for resource_type in set(resource_types):
-            try:
-                resource_install_specs.append(await self.get_code_for_type(environment, version, resource_type))
-            except CouldNotResolveCode as e:
-                LOGGER.error(
-                    "%s",
-                    e.msg,
-                )
-                invalid_resources[resource_type] = e
-            except Exception as e:
-                LOGGER.error(
-                    "Failed to get source code for %s version=%d",
-                    resource_type,
-                    version,
-                    exc_info=True,
-                )
-                invalid_resources[resource_type] = e
 
-        return resource_install_specs, invalid_resources
+        module_install_specs = []
+        stmt = (
+            select(
+                models.ModulesForAgent.module_name,
+                models.ModulesForAgent.module_version,
+                models.ModulesForAgent.cm_version,
+                models.Module.requirements,
+                models.FilesInModule.file_content_hash,
+            )
+            .join_from(
+                models.ModulesForAgent,
+                models.Module,
+                and_(
+                    models.ModulesForAgent.module_name == models.Module.name,
+                    models.ModulesForAgent.module_version == models.Module.version,
+                    models.ModulesForAgent.environment == models.Module.environment,
+                ),
+            )
+            .join_from(
+                models.Module,
```

Comment:
This could equally well join on ModulesForAgent if I understand correctly? (not that it matters)